### PR TITLE
[APIDX-810] Push generated sources to corresponding repo

### PR DESCRIPTION
**Pull request by PaulMathon**
- A step is added into each language's action, it triggers the push_sdk.sh script
- The push_sdk.sh script has been updated

After discussing with the open-source team, the use of deploy keys instead of OAuth Token was required to push SDKs for the generator to the SDK's repositories.

A deploy key is an SSH public key that can be set on a repository to clone and push from a remote server with the corresponding private key.

Pros: Access can be restricted to the repository only, more secured

Cons: If we need to clone/push on multiple repositories, being our case with one directory per language, we will need one key pair for each repo.

A Github Access Token is linked to a GitHub user, it can give access to the repositories the user has access to through HTTPS.

Pros: Only one access token is needed to clone/push on every repository, as long as the user that is linked to the token has access to it.

Cons: If the access token is leaked, the attacker will have access to the SDK repositories, but also to every other repository the linked user has access to.

As we decided to use deploy keys for obvious security reasons, we needed to set 3 private keys as environment secrets on the generator's repository: JAVA_SDK_REPO_PRIVATE_KEY, PHP_SDK_REPO_PRIVATE_KEY and PYTHON_SDK_REPO_PRIVATE_KEY. Their corresponding public key was set as deploy keys on each SDK repository.

### CODE REVIEW

**Review by ouvreboite — APPROVED** [review]

File: scripts/push_sdk.sh [hunk 1]
```diff
@@ -1,94 +1,144 @@
...
-else
-    echo "Usage: $0 BUILD_NUMBER"
-    exit 1
+ORGANIZATION_NAME="criteo"
+REPOSITORY_NAME="criteo-api-${LANGUAGE}-sdk"
+
+GENERATOR_REPO_DIR=$GITHUB_WORKSPACE
+SDK_REPO_DIR=$RUNNER_TEMP
+
+TAG_VERSION="v$GITHUB_RUN_NUMBER"
```
Comment (PaulMathon) [inline scripts/push_sdk.sh:12]: Are we sure that GITHUB_RUN_NUMBER will always be unique? (And incremental)

File: scripts/push_sdk.sh [hunk 2]
```diff
@@ -1,94 +1,144 @@
...
+  if [[ -d $sdks_directory ]]; then
+    cd $sdks_directory
+    rm -rf *
+    echo "[INFO] Success."
+  else
+    echo "[WARN] Directory $REPO_NAME/sdks doesn't exists, skipping."
+  fi
+}
+
+copy_new_sdks() {
```
Comment (bubianchi-criteo) [inline scripts/push_sdk.sh:73]: Maybe a generic remark on functions: what will be the status code if something (i.e. an internal command) has failed? 
Shouldn't we manage this status code in the caller?

File: scripts/push_sdk.sh [hunk 3]
```diff
@@ -1,94 +1,144 @@
...
+
+  if [[ ! -d $sdks_directory ]]; then
+    echo "[WARN] Directory $sdks_directory doesn't exists, creating it..."
+    mkdir $sdks_directory
+    echo "[INFO] Directory $sdks_directory created."
+  fi
+
+  cp -r "$GENERATOR_REPO_DIR/generated-sources/$LANGUAGE/." "$SDK_REPO_DIR/sdks"
+
+  echo "[INFO] Copy successful."
```
Comment (bubianchi-criteo) [inline scripts/push_sdk.sh:86]: Should we echo another message if previous "cp" failed?

File: scripts/push_sdk.sh [hunk 4]
```diff
@@ -1,94 +1,144 @@
...
+  git_clone
 
-  cd ${NEXT_CLIENT}
+  remove_previous_sdks
 
-  # add files before doing the diff
-  git_add_files ${language}
+  copy_new_sdks
+
+  git_add_files
```
Comment (bubianchi-criteo) [inline scripts/push_sdk.sh:123]: should we continue if copy_new_sdks failed (or the script will stop by itself?)

**Review by bubianchi-criteo — COMMENTED** [review]
I don't know how much my comments are relevant; but the main question is "did we manage the flow correctly and log when an error occurred?"

File: scripts/push_sdk.sh (hunk 3 above)
Comment (PaulMathon) [reply scripts/push_sdk.sh:86]: indeed we should not thanks

File: scripts/push_sdk.sh (hunk 2 above)
Comment (PaulMathon) [reply scripts/push_sdk.sh:73]: what do you mean by status code? exit code?

File: scripts/push_sdk.sh (hunk 4 above)
Comment (PaulMathon) [reply scripts/push_sdk.sh:123]: the first command line of the script, `set -ex`, should prevent this from happening, as it should stop the script if any simple command fails

File: scripts/push_sdk.sh [hunk 5]
```diff
@@ -1,94 +1,144 @@
...
+  echo "[INFO] Success. Repository cloned at $SDK_REPO_DIR"
+}
+
+remove_previous_sdks() {
+  echo "[INFO] Removing previous SDKs..."
+
+  sdks_directory="$SDK_REPO_DIR/sdks/$LANGUAGE"
+
+  if [[ -d $sdks_directory ]]; then
+    cd $sdks_directory
```
[the remark this answers is not in the document]
Comment (PaulMathon) [reply scripts/push_sdk.sh:65]: not sure to understand the question, but if a command fails, then the script should stop executing as we set `set -ex` at the beginning of the script

**Review by bubianchi-criteo — APPROVED** [review]

**Pull request merged**